org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java

```diff
@@ -0,0 +1,208 @@
+		DOMNode node = DOMDocument.findNodeOrAttrAt(xmlDocument, offset);
+		if (node.isAttribute()) {
+			DOMAttr attr = (DOMAttr) node;
+			for (DOMNode attrChild : attr.getChildren()) {
+				if (attrChild.getStart() <= offset && offset <= attrChild.getEnd()) {
+					return attrChild;
+				}
+			}
+		} else if (node.isDoctype()) {
+			DOMDocumentType doctype = (DOMDocumentType) node;
+			DTDDeclParameter subset = doctype.getInternalSubsetNode();
+			if (subset != null && subset.getStart() < offset && offset < subset.getEnd()) {
+				for (DOMNode child : doctype.getChildren()) {
+					if (child.getStart() != -1 && child.getStart() <= offset && offset < child.getEnd()) {
+						return child;
+					}
+				}
+			}
+		}
+		return node;
+	}
+
+	private SelectionRange getSelectionRangeFromNode(DOMNode node, int offset) {
+		// The leaf of the SelectionRange structure is a dummy node that doesn't
+		// represent an actual range in the document
+		final SelectionRange leafSelectionRange = new SelectionRange();
+		leafSelectionRange.setRange(new Range());
+		SelectionRange selectionRange = leafSelectionRange;
+
+		do {
+			selectionRange = handleNodeSelectionRange(selectionRange, node, offset);
+			if (node.isAttribute() && ((DOMAttr) node).getOwnerElement() != null) {
+				node = ((DOMAttr) node).getOwnerElement();
+			}
+			DOMNode prolog = node.getOwnerDocument().getProlog();
+			if (prolog != null && node != node.getOwnerDocument() && node != prolog && prolog.getStart() <= offset
+					&& offset <= prolog.getEnd()) {
+				node = node.getOwnerDocument().getProlog();
+			} else {
+				node = node.getParentNode();
+			}
+		} while (node != null);
+
+		// Skip the leaf node, which is a dummy node
+		return leafSelectionRange.getParent();
+	}
+
+	private SelectionRange handleNodeSelectionRange(SelectionRange selectionRange, DOMNode node, int offset) {
+		switch (node.getNodeType()) {
+			case Node.ELEMENT_NODE:
+				return handleElementSelectionRange(selectionRange, (DOMElement) node, offset);
+			case Node.ATTRIBUTE_NODE:
+				return handleAttributeSelectionRange(selectionRange, (DOMAttr) node, offset);
+			case Node.DOCUMENT_TYPE_NODE:
+				return handleDoctypeSelectionRange(selectionRange, (DOMDocumentType) node, offset);
+			default:
+				return handleGenericNodeSelectionRange(selectionRange, node, offset);
+		}
+	}
+
+	private SelectionRange handleAttributeSelectionRange(SelectionRange selectionRange, DOMAttr attr, int offset) {
+		selectionRange = handleGenericNodeSelectionRange(selectionRange, attr, offset);
+		DOMElement ownerElement = attr.getOwnerElement();
+		if (ownerElement != null) {
+			selectionRange = handleGenericNodeSelectionRange(selectionRange, ownerElement, offset);
+		}
+		return selectionRange;
+	}
+
+	private SelectionRange handleElementSelectionRange(SelectionRange selectionRange, DOMElement element, int offset) {
+		XMLPositionUtility.selectStartTagName(element);
```

Comment:
Does this have a side effect ? Seems to only be useful for its returned value.

Reply:
No, this won't do anything. I'll remove it.